File: src/longest_palindrome_by_concatenating_two_letter_words_2131/solution.py
from collections import Counter
from typing import List


class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        word_to_count = Counter(words)

        visited = set()
        used_center = False
        result = 0

        for word, count in word_to_count.items():
            if word in visited:
                continue

            visited.add(word)

            if word[0] == word[1]:
                if count % 2 == 0:
                    result += count * 2
                else:
                    result += (count - 1) * 2
                    if not used_center:
                        used_center = True
                        result += 2
            else:
                pair = word[1] + word[0]
                pair_count = word_to_count[pair]

                if pair_count > 0:
                    visited.add(pair)
                    result += 2 * 2 * min(count, pair_count)

        return result


def test():
    solution = Solution()
    print(solution.longestPalindrome(["lc", "cl", "gg"]))
    print(solution.longestPalindrome(["ab", "ty", "yt", "lc", "cl", "ab"]))
    print(solution.longestPalindrome(["cc", "ll", "xx"]))

File: src/longest_palindrome_by_concatenating_two_letter_words_2131/solution.rs
use std::collections::{HashMap, HashSet};

pub struct Solution;

impl Solution {
    pub fn longest_palindrome(words: Vec<String>) -> i32 {
        let mut word_to_count = HashMap::new();

        for word in &words {
            *word_to_count.entry(word.as_str()).or_insert(0) += 1;
        }

        let mut visited: HashSet<String> = HashSet::new();
        let mut used_center = false;
        let mut result = 0;

        for (&word, &count) in &word_to_count {
            if !visited.insert(word.to_string()) {
                continue;
            }

            let chars: Vec<char> = word.chars().collect();
            if chars[0] == chars[1] {
                if count % 2 == 0 {
                    result += count * 2;
                } else {
                    result += (count - 1) * 2;
                    if !used_center {
                        used_center = true;
                        result += 2;
                    }
                }
            } else {
                let pair = format!("{}{}", chars[1], chars[0]);
                if let Some(&pair_count) = word_to_count.get(pair.as_str()) {
                    visited.insert(pair);
                    result += 2 * 2 * count.min(pair_count);
                }
            }
        }

        result
    }
}

pub fn test() {
    println!(
        "{}",
        Solution::longest_palindrome(vec!["lc".to_string(), "cl".to_string(), "gg".to_string()])
    );
    println!(
        "{}",
        Solution::longest_palindrome(vec![
            "ab".to_string(),
            "ty".to_string(),
            "yt".to_string(),
            "lc".to_string(),
            "cl".to_string(),
            "ab".to_string()
        ])
    );
    println!(
        "{}",
        Solution::longest_palindrome(vec!["cc".to_string(), "ll".to_string(), "xx".to_string()])
    );
}
